@here/harp-mapview/lib/geometry/TileGeometryLoader.ts

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { DecodedTile } from "@here/harp-datasource-protocol";
+import { PerformanceTimer } from "@here/harp-utils";
+
+import { PerformanceStatistics } from "../Statistics";
+import { Tile } from "../Tile";
+import { TileGeometryCreator } from "./TileGeometryCreator";
+
+/**
+ * Loads the geometry for its [[Tile]]. Derived classes allow for different loading strategies.
+ */
+export interface TileGeometryLoader {
+    /**
+     * The [[Tile]] this `TileGeometryLoader` is managing.
+     */
+    tile: Tile;
+
+    /**
+     * `True` if all geometry of the `Tile` has been loaded and the loading process is finished.
+     */
+    isFinished: boolean;
+
+    /**
+     * `True` if the basic geometry has been loaded, and the `Tile` is ready  for display.
+     */
+    basicGeometryLoaded: boolean;
+
+    /**
+     * `True` if all geometry of the `Tile` has been loaded.
+     */
+    allGeometryLoaded: boolean;
+
+    /**
+     * Proceed with loading.
+     */
+    update(): void;
+
+    /**
+     * Dispose of any resources.
+     */
+    dispose(): void;
+}
+
+/**
+ * Simplest implementation of a [[TileGeometryLoader]]. It loads all geometry in a single step.
+ */
+export class SimpleTileGeometryLoader implements TileGeometryLoader {
+    private m_decodedTile?: DecodedTile;
+    private m_isFinished: boolean = false;
+
+    constructor(private m_tile: Tile) {}
+
+    get tile(): Tile {
+        return this.m_tile;
+    }
+
+    get isFinished(): boolean {
+        return this.m_isFinished;
+    }
+
+    get basicGeometryLoaded(): boolean {
+        return this.m_tile.hasGeometry;
+    }
+
+    get allGeometryLoaded(): boolean {
+        return this.m_isFinished;
+    }
+
+    setDecodedTile(decodedTile: DecodedTile) {
+        this.m_decodedTile = this.m_tile.decodedTile;
+    }
+
+    update(): void {
+        if (this.m_decodedTile === undefined && this.m_tile.decodedTile !== undefined) {
+            this.setDecodedTile(this.m_tile.decodedTile);
+            this.prepareForRender();
+            this.finish();
```

Comment:
Is this needed here or should it be done in timeout handler in prepareForRender?

Reply:
This makes more sense when the implementation with phased loading is visible. It already makes sense now, I think.